[IDED] Rebalances Venus Human Traps

Isaacnml opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

This PR focuses on rebalancing  aspects of the Venus Human Trap, making them more balanced to fight against

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

Makes fights less one sided against Venus Human Traps

A singular Venus Trap shouldn't be able to wipe a 20 shift pop station with a singular click, nor should it be fully chasing someone more than 10 tiles away from the vines


## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

https://github.com/user-attachments/assets/11af50a3-298c-4b26-81b6-b37e4e0b546d

https://github.com/user-attachments/assets/b50f003b-0fed-4373-954e-3593bdb44bf5

https://github.com/user-attachments/assets/7ac3444f-4046-4009-bc60-063af1bbfc38



</details>

## Changelog
:cl: Isaacnml, Gilgax
balance:  Replaces the Hardstun of Venus Human Traps with a knockdown ontop of stamina damage
balance: Cyborgs are now slowed down by 50% for 10 seconds upon being hit by  a ranged attack from the vines
balance: Venus Human Traps, Lowers their damage from 25 to 10
balance: Venus Human Traps, Health changed to 100
balance: While a mob is attached to a venus human trap, they receive 10 constant stamina damage until the vine is broken
balance: Lowered the amount of health healed by hitting a mob that is alive to 5%
balance: Venus Human Trap ranged attack is only able to hit mobs/objects
balance: Venus Human Traps only heal 5% when inside the vines
balance: Venus Human Trap AI retreats at 50% health back to the vines, regardless if they're in combat or not, provided they have line of sight with them
balance: Venus Human Trap ranged attack can be broken if something pulls the object/mob
balance: Venus Human Traps now prioritize attacking mobs inside vines 
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Isaacnml committed:
I-dead

Tsar-Salat reviewed (changes requested):
This will make the issue of ai-controlled vines killing themselves much more apparent. 

There have been several times I've had to respawn vines that randomly decided to go to non-vine turfs out of combat and die, because they weren't adjusted to consider non-vine turfs invalid since https://github.com/BeeStation/BeeStation-Hornet/pull/10533

If you want them to ignore this during combat, that is fine, but the modifier shouldn't be increased until that has been addressed.

Tsar-Salat commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -189,7 +189,7 @@
 	if(isliving(the_target))
 		var/mob/living/L = the_target
-		L.Paralyze(20)
+		L.Knockdown(10)
```
This will likely be completely useless against borgs. The vine can be often hard to aim. especially when it considers every atom a possible target that can invoke the CD. This means clicking on turfs or vines or walls can mess it up.

Lower the cooldown, add some borg behavior, and make it only be able to target objects and mobs.

I agree, hardstuns are shit, but this should be more *convenient* in exchange for being less of an instantwin

PowerfulBacon replied:
Alternatively if you click a turf then it should grab the first mob that it encounters in the line of reaching that turf

PowerfulBacon reviewed (changes requested):
Add some damage (~15) to the ranged attack and double their health to compensate. Optionally give a chance to grab and steal the primary active weapon of the player, though this may also be annoying. You have the option to making their combat more interesting, but they need to be at least somewhat of a threat.

Rukofamicom reviewed (commented):
I'm not going to place a blocking request, but I would like to suggest a complete change to how the plants operate on balance if you're willing to do it. This should make them feel more fair while still keeping them somewhat powerful in a different way. 

This is taking heavy inspiration from the tethering enemies we used in the old halloween event:

* Increase the health of the venus traps
* Remove the **absurd** melee damage the plants have and make them hit like a wet noodle on mobs. They should still be able to heavily damage terrain and objects for the sake of helping the vines spread though.
* When a vine tethers someone it should cause an immediate knockdown which lasts for a few seconds, disarming the player in the process.
* When a vine drags someone, it should refresh the knockdown so that for as long as someone is tethered, they are knocked down. Dragging should also add some amount of stamina damage (15-25) so that someone being pulled will eventually be unable to crawl away or otherwise resist being pulled into the vines.
* If someone is currently pulling the target when the tether attempts to drag them, it instead releases its victim

This keeps the venus plants threatening but less immediately lethal. They have to subdue and drag their victims into areas spaced by the vines or otherwise utilize threats imposed by the vines to finish someone off. Leaving the vines to quickly toss someone into crit should stop happening, and teamwork becomes the most important factor for both sides of the equation. 

The issue with this setup is cyborgs, as with many things. I'm not sure how to factor a solution for cyborgs into this unless we can implement a "stamina" damage that works on cyborgs too. The vines could have a check to fully restrain cyborgs while tethered from the start instead of building toward fully restrained over time? Cyborgs could be prevented from moving entirely while tethered and eventually paralyzed after a certain number of times they have been dragged?

Isaacnml committed:
Merge branch 'BeeStation:master' into Ided]-Nerfs-Venus-Human-Trap

Isaacnml committed:
General Changes

Isaacnml committed:
Revert "General Changes"

This reverts commit 3aaf80ca4e6b537b61bc983e9798859e622291aa.

Isaacnml committed:
General Changes 2

Hopefully i didn't fuck up

Gilgaxx committed:
Wisty's commit

Added a check for "if no ckey, health below 50% and outside of vines, it makes a list of all possible turfs with vines, loses all targets, stops attacking and goes to a random tile with vine.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -184,17 +184,29 @@
+	var/mob/living/L = the_target
+	if(iscarbon(L))
+		L.apply_damage(30, STAMINA, BODY_ZONE_CHEST)
+		L.Knockdown(1 SECONDS)
```
```suggestion
		L.Knockdown(3 SECONDS)
```

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -234,12 +246,20 @@
+
+			var/mob/living/L = B.target
+			L.apply_damage(5, STAMINA, BODY_ZONE_CHEST)
+			to_chat(L, span_userdanger("Your body feels weaker!"))
```
```suggestion
			L.Knockdown(3 SECONDS)
```

This message will spam, and the knockdown needs to be refreshed or else they will easily walk away from the vine and break it after the 1 second is up.

This forced them to crawl instead of running, which may still enable them to get far enough to break

Rukofamicom reviewed (changes requested):
Tweaks I will request since you went along with my suggestion. These are necessary for them to still present an actual threat with the reduced damage.

This isn't quite what I had requested but will be functional otherwise and looks good. The code for making a vine break if being pulled isn't there, but the stamina damage is also lower so this should work out fine. The vine will still break if someone is pulled far enough away and this adds more tension to it.

Isaacnml committed:
Merge pull request Wisty

Wistys PR for retreating Venus man traps SECOND ATTEMPT

Isaacnml committed:
Ruko Suggestions 1

Co-authored-by: Rukofamicom <oozelingcomplaints@gmail.com>

Rukofamicom reviewed (approved):
Looks good to me now

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -184,17 +202,29 @@
+	if(iscarbon(L))
+		L.apply_damage(30, STAMINA, BODY_ZONE_CHEST)
+		L.Knockdown(3 SECONDS)
+		to_chat(L, span_alert("The vines knock you down"))
```
```suggestion
		L.apply_damage(30, STAMINA, BODY_ZONE_CHEST)
		L.Knockdown(3 SECONDS)
		L?.pulledby.stop_pulling()
		to_chat(L, span_alert("The vines knock you down"))
```

This will cause a vines to break grabs so players will have to grab the person again. If this wasn't here, players could simply hold each other to be immune to vines grabbing at all

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -234,12 +264,20 @@
 		if(istype(B.target, /atom/movable))
 			var/atom/movable/AM = B.target
 			if(!AM.anchored)
 				step(AM, get_dir(AM,src))
```
```suggestion

/mob/living/simple_animal/hostile/venus_human_trap/proc/pull_vines()
	for(var/datum/beam/B in vines)
		if(istype(B.target, /atom/movable))
			var/atom/movable/AM = B.target
			if(AM.pulledby) //If someone is pulling the victim that a vine has hit, release them.
				qdel(B)
				return //This vine no longer exists abort
			if(!AM.anchored)
				step(AM, get_dir(AM,src))
```

This will cause the vines to break instead of pulling if they try to pull while someone is already pulling their victim.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -234,12 +264,20 @@
+		if(iscarbon(B.target)) // If they dont get away quickly, make them take constant stamina damage
+
+			var/mob/living/L = B.target
+			L.apply_damage(5, STAMINA, BODY_ZONE_CHEST)
```
```suggestion
			L.apply_damage(20, STAMINA, BODY_ZONE_CHEST)
```

With players possible to easily rescue this needs to be higher as originally suggested so that they are less able to crawl away without help

Rukofamicom reviewed (commented):
Here's the code for making the vines break when someone is pulling. 

Taken from my own PR here: https://github.com/PestoVerde322/BeeStation-Hornet/pull/8

Isaacnml committed:
Ruko Suggestions 2

Co-authored-by: Rukofamicom <oozelingcomplaints@gmail.com>

Isaacnml committed:
Last changes

Isaacnml committed:
Whoops

Isaacnml committed:
Revert "Whoops"

This reverts commit f02efd31d43f1daba98334c2e13ca96b37903155.

Isaacnml committed:
Fix

Isaacnml committed:
Indent Fix

PowerfulBacon reviewed (approved):
Excellent work

Tsar-Salat reviewed (approved):
We made a good ided out of you :>

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -228,29 +259,45 @@
+
+		if(iscarbon(B.target)) // If they dont get away quickly, make them take constant stamina damage
+			var/mob/living/L = B.target
+			L.apply_damage(20, STAMINA, BODY_ZONE_CHEST)
```
```suggestion
			L.apply_damage(10, STAMINA, BODY_ZONE_CHEST)
```

As it turns out, this was too much I think!

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -184,22 +201,36 @@
+
+	var/mob/living/L = the_target
+	if(iscarbon(L))
+		L.apply_damage(30, STAMINA, BODY_ZONE_CHEST)
```
```suggestion
		L.apply_damage(20, STAMINA, BODY_ZONE_CHEST)
```

This too!

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm:
```diff
@@ -149,25 +149,42 @@
 			to_chat(src, span_notice(" The vines nourish you, healing your wounds."))
-		adjustHealth(-maxHealth*0.05)
+			stop_automated_movement = 0
+		adjustHealth(-maxHealth*0.1)
```
```suggestion
		adjustHealth(-maxHealth*0.05)
```

I don't believe they should heal as fast as they do now that they have this much health having observed and participating in some hit and run antics. Being able to soak directly through a cyborg attacking us while we were in the vines was a bit much.

Rukofamicom reviewed (changes requested):
Suggested tweaks since they performed a bit better than expected against a relatively coordinated crew.

Isaacnml committed:
Nerfs

Co-authored-by: Rukofamicom <oozelingcomplaints@gmail.com>

Isaacnml committed:
More fixes

Isaacnml committed:
Take 2
